### PR TITLE
no_std-support

### PR DESCRIPTION
**Pull request by LucasLvy**
We created 2 new crates to break some dependencies and have less crates to turn no_std. Note that this PR introduces no_std compatibility for :
* `cairo-lang-casm`
* `cairo-lang-casm-contract-class` 
* `cairo-lang-utils` 
* `cairo-lang-vm-utils`

Introduction of a serde feature for all the crates that use serde.

This change is [Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/3771)

### CODE REVIEW

File: ensure_no_std/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+version = "0.1.0"
+edition = "2021"
+
+[dependencies]
+cairo-lang-utils = { path = "../crates/cairo-lang-utils", default-features = false }
+cairo-lang-casm = { path = "../crates/cairo-lang-casm", default-features = false }
+cairo-lang-vm-utils = { path = "../crates/cairo-lang-vm-utils", default-features = false, features = [
+    "execute_core_hints",
+] }
+cairo-lang-casm-contract-class = { path = "../crates/cairo-lang-casm-contract-class", default-features = false }
```
Comment (tdelabro) [inline ensure_no_std/Cargo.toml:12]: You can add the  `features = ["serde"]` and any other feature that exists and should be supported in `no_std`

**Review by tdelabro — CHANGES_REQUESTED** [review]

File: crates/cairo-lang-runner/Cargo.toml [hunk 2]
```diff
@@ -25,16 +27,21 @@ cairo-lang-sierra-ap-change = { path = "../cairo-lang-sierra-ap-change", version
...
+    "execute_core_hints",
+] }
+cairo-lang-sierra-type-size = { path = "../cairo-lang-sierra-type-size", version = "2.1.0-rc0" }
 cairo-vm.workspace = true
 itertools.workspace = true
 keccak.workspace = true
 num-bigint.workspace = true
 num-integer.workspace = true
 num-traits.workspace = true
-salsa.workspace = true
```
[the remark this answers is not in the document]
Comment (LucasLvy) [reply crates/cairo-lang-runner/Cargo.toml:37]: apparently as everything still works

File: crates/cairo-lang-vm-utils/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+mod my_macro {
+    /// Inserts a value into the vm memory cell represented by the cellref.
+    #[macro_export]
+    macro_rules! insert_value_to_cellref {
+        ($vm:ident, $cell_ref:ident, $value:expr) => {
+            $vm.insert_value($crate::cell_ref_to_relocatable($cell_ref, $vm), $value)
+        };
+    }
+}
+
```
[the remark this answers is not in the document]
Comment (LucasLvy) [reply crates/cairo-lang-vm-utils/src/lib.rs:40]: yes

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -39,8 +47,18 @@ jobs:
...
           - nextest run -p cairo-lang-utils
+          - nextest run -p cairo-lang-utils --no-default-features
+          - nextest run -p cairo-lang-utils --no-default-features --features parity-scale-codec
+          - nextest run -p cairo-lang-utils --no-default-features --features serde
+          - nextest run -p cairo-lang-utils --no-default-features --all-features
+          - nextest run -p cairo-lang-vm-utils
+          - nextest run -p cairo-lang-vm-utils --no-default-features
+          - nextest run -p cairo-lang-casm-contract-class
+          - nextest run -p cairo-lang-casm-contract-class --no-default-features
+          - nextest run -p cairo-lang-casm-contract-class --no-default-features --features serde
```
Comment (tdelabro) [inline .github/workflows/ci.yml:61]: Is all this really required?
Can't you just do a run with no_std + supported features and another with all features including std?

File: crates/cairo-lang-casm/src/operand.rs [hunk 5]
```diff
@@ -65,13 +70,17 @@ impl<T: Into<BigIntAsHex>> From<T> for ResOperand {
 }
 
 /// Represents an operand of the form [reg + offset].
-#[derive(Serialize, Deserialize, Copy, Clone, Debug, Eq, PartialEq, Encode, Decode, JsonSchema)]
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[cfg_attr(feature = "schemars", derive(schemars::JsonSchema))]
+#[cfg_attr(feature = "parity-scale-codec", derive(Encode, Decode))]
+
```
Comment (tdelabro) [inline crates/cairo-lang-casm/src/operand.rs:77]: remove empty line

File: crates/cairo-lang-filesystem/Cargo.toml [hunk 6]
```diff
@@ -11,10 +11,13 @@ cairo-lang-debug = { path = "../cairo-lang-debug", version = "2.1.0-rc0" }
 cairo-lang-utils = { path = "../cairo-lang-utils", version = "2.1.0-rc0" }
 path-clean.workspace = true
 salsa.workspace = true
-serde.workspace = true
+serde = { workspace = true, optional = true }
```
Comment (tdelabro) [inline crates/cairo-lang-filesystem/Cargo.toml:14]: It should also be imported with feature `std`.

File: crates/cairo-lang-runner/Cargo.toml [hunk 7]
```diff
@@ -13,7 +13,9 @@ ark-secp256k1.workspace = true
 ark-secp256r1.workspace = true
 ark-std.workspace = true
 cairo-felt.workspace = true
```
Comment (tdelabro) [inline crates/cairo-lang-runner/Cargo.toml:15]: features = std

File: crates/cairo-lang-runner/Cargo.toml [hunk 8]
```diff
@@ -25,16 +27,21 @@ cairo-lang-sierra-ap-change = { path = "../cairo-lang-sierra-ap-change", version
...
+    "std",
+    "execute_core_hints",
+] }
+cairo-lang-sierra-type-size = { path = "../cairo-lang-sierra-type-size", version = "2.1.0-rc0" }
 cairo-vm.workspace = true
 itertools.workspace = true
 keccak.workspace = true
 num-bigint.workspace = true
 num-integer.workspace = true
 num-traits.workspace = true
```
Comment (tdelabro) [inline crates/cairo-lang-runner/Cargo.toml:44]: For all those too feature = "std"

**Review by tdelabro — CHANGES_REQUESTED** [review]
When a crate does not aim to build in `no_std`. It should import the deps defined in the workspace with `default-features = false`, using `features = ["std"]`. 
So it can benefit from those potential additional features

**Review by orizi — DISMISSED** [review]
> Reviewed 16 of 59 files at r1, 6 of 51 files at r2, 5 of 31 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/3771)* status: 27 of 98 files reviewed, 15 unresolved discussions (waiting on @LucasLvy, @spapinistarkware, and @tdelabro)

___
*[`.gitignore` line 11 at r3](https://reviewable.io/reviews/starkware-libs/cairo/3771#-NaLSawW4xqAn4wHzSdR:-NaLSawW4xqAn4wHzSdS:undefined) ([raw file](https://github.com/starkware-libs/cairo/blob/3142b5ffba7b7e2204776661dbb44e6787a1a54c/.gitignore#L11)):*
> ```ignore
> # so it is kept out of the workspace meaning it produce it's own target and Cairo.lock files
> ensure_no_std/target
> ensure_no_std/Cargo.lock
> ```





_Suggestion:_
```ignore
# This crate has to be built with nightly and specific target.
# so it is kept out of the workspace meaning it produce it's own target and Cairo.lock files
ensure_no_std/target
ensure_no_std/Cargo.lock
```

**Review by LucasLvy — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/3771)* status: 26 of 98 files reviewed, 15 unresolved discussions (waiting on @orizi, @spapinistarkware, and @tdelabro)

___
*[`.gitignore` line 11 at r3](https://reviewable.io/reviews/starkware-libs/cairo/3771#-NaLSawW4xqAn4wHzSdR:-NaLWY7C0PC8D8Qa2z3Y:b-896fix) ([raw file](https://github.com/starkware-libs/cairo/blob/3142b5ffba7b7e2204776661dbb44e6787a1a54c/.gitignore#L11)):*
> ```ignore
> # so it is kept out of the workspace meaning it produce it's own target and Cairo.lock files
> ensure_no_std/target
> ensure_no_std/Cargo.lock
> ```

Done.
___
*[`.github/workflows/ci.yml` line 61 at r2](https://reviewable.io/reviews/starkware-libs/cairo/3771#gh-1275068476:-NaLWYvT9G93WRWoK7bP:b-896fix) ([raw file](https://github.com/starkware-libs/cairo/blob/afee58b48e9d3b244bcb56c1b9ef5631abf65795/.github/workflows/ci.yml#L61)):*
<details><summary><i>Previously, tdelabro (Timothée Delabrouille) wrote…</i></summary><blockquote>

Is all this really required?
Can't you just do a run with no_std + supported features and another with all features including std?
</blockquote></details>

Done.
___
*[`crates/cairo-lang-casm/Cargo.toml` line 34 at r1](https://reviewable.io/reviews/starkware-libs/cairo/3771#gh-1274151008:-NaLWZFvAGad7W8ySuNk:b-896fix) ([raw file](https://github.com/starkware-libs/cairo/blob/45a7ee8868529e3bf836d2eef90e4042aac4d358/crates/cairo-lang-casm/Cargo.toml#L34)):*
<details><summary><i>Previously, tdelabro (Timothée Delabrouille) wrote…</i></summary><blockquote>

also `thiserror-no-std/std`, `parity-scale-codec?/std`, `cairo-lang-utils/std`
</blockquote></details>

Done.
___
*[`crates/cairo-lang-casm/src/operand.rs` line 77 at r2](https://reviewable.io/reviews/starkware-libs/cairo/3771#gh-1275072816:-NaLWZZh3PrDVIiPxNiY:b-896fix) ([raw file](https://github.com/starkware-libs/cairo/blob/afee58b48e9d3b244bcb56c1b9ef5631abf65795/crates/cairo-lang-casm/src/operand.rs#L77)):*
<details><summary><i>Previously, tdelabro (Timothée Delabrouille) wrote…</i></summary><blockquote>

remove empty line
</blockquote></details>

Done.
___
*[`crates/cairo-lang-casm-contract-class/Cargo.toml` line 21 at r1](https://reviewable.io/reviews/starkware-libs/cairo/3771#gh-1274149078:-NaLW_-zBSvrUgRhIK-H:b-896fix) ([raw file](https://github.com/starkware-libs/cairo/blob/45a7ee8868529e3bf836d2eef90e4042aac4d358/crates/cairo-lang-casm-contract-class/Cargo.toml#L21)):*
<details><summary><i>Previously, tdelabro (Timothée Delabrouille) wrote…</i></summary><blockquote>

itertools has a `use_std` feature
</blockquote></details>

Done.
___
*[`crates/cairo-lang-casm-contract-class/src/lib.rs` line 12 at r1](https://reviewable.io/reviews/starkware-libs/cairo/3771#gh-1274149748:-NaLW_E07oOWY34dOvnp:b-896fix) ([raw file](https://github.com/starkware-libs/cairo/blob/45a7ee8868529e3bf836d2eef90e4042aac4d358/crates/cairo-lang-casm-contract-class/src/lib.rs#L12)):*
<details><summary><i>Previously, tdelabro (Timothée Delabrouille) wrote…</i></summary><blockquote>

Everything serde related could be behind a feature flag
</blockquote></details>

Done.
___
*[`crates/cairo-lang-filesystem/Cargo.toml` line 14 at r2](https://reviewable.io/reviews/starkware-libs/cairo/3771#gh-1275079160:-NaLW_TS5A5j3rpi7ydb:b-896fix) ([raw file](https://github.com/starkware-libs/cairo/blob/afee58b48e9d3b244bcb56c1b9ef5631abf65795/crates/cairo-lang-filesystem/Cargo.toml#L14)):*
<details><summary><i>Previously, tdelabro (Timothée Delabrouille) wrote…</i></summary><blockquote>

It should also be imported with feature `std`.
</blockquote></details>

Done.
___
*[`crates/cairo-lang-project/Cargo.toml` line 12 at r2](https://reviewable.io/reviews/starkware-libs/cairo/3771#gh-1275077122:-NaLW_mkE173Ibn2g6DZ:b-896fix) ([raw file](https://github.com/starkware-libs/cairo/blob/afee58b48e9d3b244bcb56c1b9ef5631abf65795/crates/cairo-lang-project/Cargo.toml#L12)):*
<details><summary><i>Previously, tdelabro (Timothée Delabrouille) wrote…</i></summary><blockquote>

it should also be imported with "std" feature, as this crate does not aim to compile in `no_std`.
</blockquote></details>

Done.
___
*[`crates/cairo-lang-runner/Cargo.toml` line 15 at r2](https://reviewable.io/reviews/starkware-libs/cairo/3771#gh-1275080551:-NaLWaFT9s7ZoINC8SIc:b-896fix) ([raw file](https://github.com/starkware-libs/cairo/blob/afee58b48e9d3b244bcb56c1b9ef5631abf65795/crates/cairo-lang-runner/Cargo.toml#L15)):*
<details><summary><i>Previously, tdelabro (Timothée Delabrouille) wrote…</i></summary><blockquote>

features = std
</blockquote></details>

Done.
___
*[`crates/cairo-lang-runner/Cargo.toml` line 44 at r2](https://reviewable.io/reviews/starkware-libs/cairo/3771#gh-1275081081:-NaLWaU0-Ty6B0lr_Kgs:b-896fix) ([raw file](https://github.com/starkware-libs/cairo/blob/afee58b48e9d3b244bcb56c1b9ef5631abf65795/crates/cairo-lang-runner/Cargo.toml#L44)):*
<details><summary><i>Previously, tdelabro (Timothée Delabrouille) wrote…</i></summary><blockquote>

For all those too feature = "std"
</blockquote></details>

Done.
___
*[`crates/cairo-lang-utils/Cargo.toml` line 0 at r1](https://reviewable.io/reviews/starkware-libs/cairo/3771#gh-1274157565:-NaLWame7JlD-tpZYzzk:b-896fix) ([raw file](https://github.com/starkware-libs/cairo/blob/45a7ee8868529e3bf836d2eef90e4042aac4d358/crates/cairo-lang-utils/Cargo.toml#L0)):*
<details><summary><i>Previously, tdelabro (Timothée Delabrouille) wrote…</i></summary><blockquote>

*(Reviewable was [unable](https://github.com/Reviewable/Reviewable/issues/14) to map this GitHub inline comment thread to the right spot — sorry!)*

`schemars = ["std", "dep:shemars"]`
</blockquote></details>

Done.
___
*[`crates/cairo-lang-utils/Cargo.toml` line 38 at r1](https://reviewable.io/reviews/starkware-libs/cairo/3771#gh-1274157037:-NaLWbCoCiRjuAROxPpH:b-896fix) ([raw file](https://github.com/starkware-libs/cairo/blob/45a7ee8868529e3bf836d2eef90e4042aac4d358/crates/cairo-lang-utils/Cargo.toml#L38)):*
<details><summary><i>Previously, tdelabro (Timothée Delabrouille) wrote…</i></summary><blockquote>

`parity-scale-codec?/std`
</blockquote></details>

Done.
___
*[`ensure_no_std/Cargo.toml` line 12 at r1](https://reviewable.io/reviews/starkware-libs/cairo/3771#gh-1274161962:-NaLWbwa5gJxAvjefJFm:b-896fix) ([raw file](https://github.com/starkware-libs/cairo/blob/45a7ee8868529e3bf836d2eef90e4042aac4d358/ensure_no_std/Cargo.toml#L12)):*
<details><summary><i>Previously, tdelabro (Timothée Delabrouille) wrote…</i></summary><blockquote>

You can add the  `features = ["serde"]` and any other feature that exists and should be supported in `no_std`
</blockquote></details>

Done.

**Comment by tdelabro** [comment]
@orizi @spapinistarkware 
Do you plan to merge this one? Like, if @LucasLvy rebase it tomorrow. Will you review and merge it?

**Pull request closed without merging**